File: agent/src/registry/ref.py
"""Strict ``registry://`` reference grammar for the agent asset registry (#246).

::

    registry://<kind>/<namespace>/<name>@<constraint>
      kind       = [a-z][a-z0-9_]*            snake_case: mcp_server, cedar_policy_module
      namespace  = [a-z][a-z0-9-]*
      name       = [a-z0-9][a-z0-9._-]*
      constraint = [^~]?MAJOR.MINOR.PATCH[-prerelease]   exact / caret / tilde only

The ``@<constraint>`` pin is MANDATORY (fail-closed: no implicit "latest").

This module mirrors ``cdk/src/handlers/shared/registry/ref.ts`` byte-for-byte and
is exercised by the ``contracts/registry-resolution/`` parity corpus. Keep the two
in lockstep — a change here without the matching TS change (or vice versa) is a
parity break that CI must catch.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# MVP asset kinds the registry loads end-to-end or stages.
REGISTRY_KINDS = ("mcp_server", "cedar_policy_module", "skill")
# Reserved kinds accepted by the grammar but rejected at publish (no loader yet).
RESERVED_KINDS = ("plugin", "subagent", "prompt_fragment", "capability")

# Structural split — scheme + 3 path segments + the (mandatory) constraint.
_REF_SHAPE = re.compile(
    r"^registry://([a-z][a-z0-9_]*)/([a-z][a-z0-9-]*)/([a-z0-9][a-z0-9._-]*)@(.+)$"
)
# exact / caret / tilde over MAJOR.MINOR.PATCH with an optional prerelease.
# Rejects ``*``, ``latest``, ``>=``, ``<=``, x-ranges, and bare prerelease modifiers.
_CONSTRAINT = re.compile(
    r"^([\^~]?)(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-([0-9A-Za-z.-]+))?$"
)
_OP_BY_PREFIX = {"": "exact", "^": "caret", "~": "tilde"}


class RefError(ValueError):
    """Raised when a ``registry://`` ref is malformed.

    ``reason`` is one of ``INVALID_REGISTRY_REF`` / ``INVALID_CONSTRAINT`` — the
    same reason tokens the TS resolver reports, so both sides agree on *why* a
    ref failed, not just *that* it failed.
    """

    def __init__(self, reason: str, message: str) -> None:
        super().__init__(message)
        self.reason = reason


@dataclass(frozen=True)
class ParsedConstraint:
    op: str  # exact | caret | tilde
    major: int
    minor: int
    patch: int
    prerelease: str | None
    raw: str


@dataclass(frozen=True)
class ParsedRef:
    kind: str
    namespace: str
    name: str
    constraint: ParsedConstraint


def parse_constraint(raw: str) -> ParsedConstraint | None:
    """Parse + validate a constraint string in isolation. ``None`` if invalid."""
    m = _CONSTRAINT.match(raw)
    if not m:
        return None
    prefix, major, minor, patch, prerelease = m.groups()
    return ParsedConstraint(
        op=_OP_BY_PREFIX[prefix],
        major=int(major),
        minor=int(minor),
        patch=int(patch),
        prerelease=prerelease,
        raw=raw,
    )


def parse_ref(ref: str) -> ParsedRef:
    """Parse a strict ``registry://kind/namespace/name@constraint`` reference.

    Raises ``RefError`` (with a ``reason``) on a malformed ref or a floating /
    unsupported constraint — pins are mandatory.
    """
    shape = _REF_SHAPE.match(ref)
    if not shape:
        raise RefError(
            "INVALID_REGISTRY_REF",
            f"not a valid registry ref (expected registry://kind/namespace/name@constraint): {ref}",
        )
    kind, namespace, name, raw_constraint = shape.groups()
    constraint = parse_constraint(raw_constraint)
    if constraint is None:
        raise RefError(
            "INVALID_CONSTRAINT",
            f"unsupported version constraint '{raw_constraint}' (use exact, ^, or ~)",
        )
    return ParsedRef(kind=kind, namespace=namespace, name=name, constraint=constraint)

File: cdk/src/handlers/shared/registry/ref.ts
/**
 *  MIT No Attribution
 *
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy of
 *  the Software without restriction, including without limitation the rights to
 *  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 *  the Software, and to permit persons to whom the Software is furnished to do so.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 */

// Strict `registry://` reference grammar for the agent asset registry (#246).
//
//   registry://<kind>/<namespace>/<name>@<constraint>
//     kind       = [a-z][a-z0-9_]*            snake_case: mcp_server, cedar_policy_module
//     namespace  = [a-z][a-z0-9-]*
//     name       = [a-z0-9][a-z0-9._-]*
//     constraint = [^~]?MAJOR.MINOR.PATCH[-prerelease]   exact / caret / tilde only
//
// The `@<constraint>` pin is MANDATORY (fail-closed: no implicit "latest").
// This grammar is mirrored byte-for-byte by `parse_ref` in
// agent/src/registry/ref.py and exercised by the contracts/registry-resolution/
// parity corpus. Keep the two in lockstep.

/** MVP asset kinds the registry loads end-to-end or stages. */
export const REGISTRY_KINDS = [
  'mcp_server',
  'cedar_policy_module',
  'skill',
] as const;
export type RegistryKind = (typeof REGISTRY_KINDS)[number];

/** Reserved kinds accepted by the grammar but rejected at publish (no loader yet). */
export const RESERVED_KINDS = ['plugin', 'subagent', 'prompt_fragment', 'capability'] as const;

export type ConstraintOp = 'exact' | 'caret' | 'tilde';

export interface ParsedConstraint {
  readonly op: ConstraintOp;
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
  /** Prerelease tag without the leading `-`, or undefined. */
  readonly prerelease?: string;
  /** The constraint exactly as written (e.g. `^1.4.1`). */
  readonly raw: string;
}

export interface ParsedRef {
  readonly kind: string;
  readonly namespace: string;
  readonly name: string;
  readonly constraint: ParsedConstraint;
}

export type RefErrorReason = 'INVALID_REGISTRY_REF' | 'INVALID_CONSTRAINT';

export type ParseResult =
  | { readonly ok: true; readonly ref: ParsedRef }
  | { readonly ok: false; readonly reason: RefErrorReason; readonly message: string };

// Structural split — validates the scheme + 3 path segments and captures the
// (mandatory) constraint. Segment character classes are validated separately so
// we can distinguish a bad ref shape from a bad constraint.
const REF_SHAPE =
  /^registry:\/\/([a-z][a-z0-9_]*)\/([a-z][a-z0-9-]*)\/([a-z0-9][a-z0-9._-]*)@(.+)$/;

// exact / caret / tilde over MAJOR.MINOR.PATCH with an optional prerelease.
// Rejects `*`, `latest`, `>=`, `<=`, x-ranges, and bare prerelease modifiers.
const CONSTRAINT =
  /^([\^~]?)(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-([0-9A-Za-z.-]+))?$/;

const OP_BY_PREFIX: Record<string, ConstraintOp> = { '': 'exact', '^': 'caret', '~': 'tilde' };

/** Parse + validate a constraint string in isolation (exported for the resolver). */
export function parseConstraint(raw: string): ParsedConstraint | null {
  const m = CONSTRAINT.exec(raw);
  if (!m) return null;
  const major = Number(m[2]);
  const minor = Number(m[3]);
  const patch = Number(m[4]);
  // Reject components beyond MAX_SAFE_INTEGER: `Number()` rounds them, so TS
  // would silently compare a different value than Python's arbitrary-precision
  // int — a cross-language parity break in version selection (#246 review).
  if (!Number.isSafeInteger(major) || !Number.isSafeInteger(minor) || !Number.isSafeInteger(patch)) {
    return null;
  }
  return {
    op: OP_BY_PREFIX[m[1]],
    major,
    minor,
    patch,
    prerelease: m[5],
    raw,
  };
}

/**
 * Parse a strict `registry://kind/namespace/name@constraint` reference.
 * A ref with no `@constraint`, or a floating constraint (`*`, `latest`, `>=`…),
 * is rejected — pins are mandatory.
 */
export function parseRef(ref: string): ParseResult {
  const shape = REF_SHAPE.exec(ref);
  if (!shape) {
    return {
      ok: false,
      reason: 'INVALID_REGISTRY_REF',
      message: `not a valid registry ref (expected registry://kind/namespace/name@constraint): ${ref}`,
    };
  }
  const [, kind, namespace, name, rawConstraint] = shape;
  const constraint = parseConstraint(rawConstraint);
  if (!constraint) {
    return {
      ok: false,
      reason: 'INVALID_CONSTRAINT',
      message: `unsupported version constraint '${rawConstraint}' (use exact, ^, or ~)`,
    };
  }
  return { ok: true, ref: { kind, namespace, name, constraint } };
}
